sdks/python/apache_beam/runners/worker/worker_status.py

```diff
@@ -0,0 +1,139 @@
+    stack_trace = ''.join(traceback.format_stack(frames[t.ident]))
+    thread_ident_name = (t.ident, t.name)
+    stack_traces[stack_trace].append(thread_ident_name)
+
+  all_traces = ['=' * 10 + 'THREAD DUMP' + '=' * 10]
+  for stack, identity in stack_traces.items():
+    ident, name = identity[0]
+    trace = '--- Thread #%s name: %s %s---\n' % (
+        ident, name, 'and other %d threads' %
+        (len(identity) - 1) if len(identity) > 1 else '')
+    if len(identity) > 1:
+      trace += 'threads: %s\n' % identity
+    trace += stack
+    all_traces.append(trace)
+  all_traces.append('=' * 30)
+  return '\n'.join(x.encode('utf-8') for x in all_traces)
+
+
+def active_processing_bundles_state(bundle_process_cache):
+  active_bundles = ['=' * 10 + 'ACTIVE PROCESSING BUNDLES' + '=' * 10]
+  if not bundle_process_cache.active_bundle_processors:
+    active_bundles.append("No active processing bundles.")
+  else:
+    cache = []
+    for instruction in list(
+        bundle_process_cache.active_bundle_processors.keys()):
+      processor = bundle_process_cache.lookup(instruction)
+      if processor:
+        info = processor.state_sampler.get_info()
+        cache.append((instruction,
+                      processor.process_bundle_descriptor.id,
+                      info.tracked_thread, info.time_since_transition))
+    # reverse sort active bundle by time since last transition, keep top 10.
+    cache.sort(key=lambda x: x[-1], reverse=True)
+    for s in cache[:10]:
+      state = '--- instruction %s ---\n' % s[0]
+      state += 'ProcessBundleDescriptorId: %s\n' % s[1]
+      state += "tracked thread: %s\n" % s[2]
+      state += "time since transition: %.2f seconds\n" % (s[3] / 1e9)
+      active_bundles.append(state)
+
+  active_bundles.append('=' * 30)
+  return '\n'.join(x.encode('utf-8') for x in active_bundles)
+
+
+DONE = object()
+
+
+class FnApiWorkerStatusHandler(object):
+  def __init__(self, status_address, bundle_process_cache=None):
+    self._alive = True
+    self._bundle_process_cache = bundle_process_cache
+    ch = GRPCChannelFactory.insecure_channel(status_address)
+    grpc.channel_ready_future(ch).result(timeout=60)
+    self._status_channel = grpc.intercept_channel(ch, WorkerIdInterceptor())
+    self._status_stub = beam_fn_api_pb2_grpc.BeamFnWorkerStatusStub(
+        self._status_channel)
+    self._responses = queue.Queue()
+    self._server = threading.Thread(target=lambda: self._serve(),
+                                    name='fn_api_status_handler')
+    self._server.daemon = True
+    self._server.start()
+
+  def _get_responses(self):
+    while True:
+      response = self._responses.get()
+      if response is DONE:
+        self._alive = False
+        return
+      yield response
+
+  def _serve(self):
+    while self._alive:
+      for request in self._status_stub.WorkerStatus(self._get_responses()):
+        try:
+          response = self.generate_status_response()
```

Comment:
We can move the response collection here and avoid continue later on to make code simple.

```suggestion
          self._responses.put(beam_fn_api_pb2.WorkerStatusResponse(id=request.id, status_info=self.generate_status_response()))
```

Reply:
done.